Bottom: a33fef7511c654e5c01fbd1fd3f1d5d73c5d70c4
Top:    d3deb4215713340589dd8fb59417ea849db0262d
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-08-24 00:48:12 +0100

Fix grid extents in 3D view


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 3bc518c..837d0d5 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -53,7 +53,7 @@ static GLfloat last_modelview_matrix[4][4] = {{1.0, 0.0, 0.0, 0.0},
                                               {0.0, 0.0, 0.0, 1.0}};
 static int global_view_2d = 1;
 
-static void ghid_draw_grid ();
+static void ghid_draw_grid (BoxTypePtr);
 
 struct render_priv {
   GdkGLConfig *glconfig;
@@ -303,7 +303,7 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
   gui->set_layer (NULL, GetLayerGroupNumberByNumber (INDEXOFCURRENT), 0);
   gui->set_layer (NULL, SL_FINISHED, 0);
 
-  ghid_draw_grid ();
+  ghid_draw_grid (&region);
 
   hidgl_init_triangle_array (&buffer);
   ghid_invalidate_current_gc ();
@@ -611,7 +611,7 @@ ghid_make_gc (void)
 }
 
 static void
-ghid_draw_grid ()
+ghid_draw_grid (BoxTypePtr drawn_area)
 {
   static GLfloat *points = 0;
   static int npoints = 0;
@@ -640,10 +640,10 @@ ghid_draw_grid ()
              gport->grid_color.green / 65535.,
              gport->grid_color.blue / 65535.);
 
-  x1 = GRIDFIT_X (SIDE_X (gport->view_x0), PCB->Grid);
-  y1 = GRIDFIT_Y (SIDE_Y (gport->view_y0), PCB->Grid);
-  x2 = GRIDFIT_X (SIDE_X (gport->view_x0 + gport->view_width - 1), PCB->Grid);
-  y2 = GRIDFIT_Y (SIDE_Y (gport->view_y0 + gport->view_height - 1), PCB->Grid);
+  x1 = GRIDFIT_X (MAX (0, drawn_area->X1), PCB->Grid);
+  y1 = GRIDFIT_Y (MAX (0, drawn_area->Y1), PCB->Grid);
+  x2 = GRIDFIT_X (MIN (PCB->MaxWidth, drawn_area->X2), PCB->Grid);
+  y2 = GRIDFIT_Y (MIN (PCB->MaxHeight, drawn_area->Y2), PCB->Grid);
   if (x1 > x2)
     {
       int tmp = x1;
